### PR TITLE
BUG: Fix+test dataframe tranpose with datetimeTZ

### PR DESCRIPTION
**Pull request by jbrockmendel**
Reminder of the bug here:

```
>>> dti = pd.date_range('1977-04-15', periods=3, freq='MS', tz='US/Hawaii')
>>> df = pd.DataFrame(dti)
>>> df.T
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/frame.py", line 2562, in transpose
    return super(DataFrame, self).transpose(1, 0, **kwargs)
  File "pandas/core/generic.py", line 686, in transpose
    new_values = self.values.transpose(axes_numbers)
  File "pandas/core/base.py", line 672, in transpose
    nv.validate_transpose(args, kwargs)
  File "pandas/compat/numpy/function.py", line 56, in __call__
    self.defaults)
  File "pandas/util/_validators.py", line 218, in validate_args_and_kwargs
    validate_kwargs(fname, kwargs, compat_args)
  File "pandas/util/_validators.py", line 157, in validate_kwargs
    _check_for_default_values(fname, kwds, compat_args)
  File "pandas/util/_validators.py", line 69, in _check_for_default_values
    format(fname=fname, arg=key)))
ValueError: the 'axes' parameter is not supported in the pandas implementation of transpose()
```

The implementation is an unholy mess.  It can be made a little bit prettier, but to actually get a nice implementation we need to allow 2D EAs.

### CODE REVIEW

File: pandas/tests/frame/test_reshape.py [hunk 1]
```diff
@@ -939,3 +939,34 @@ def test_unstack_fill_frame_object():
         index=list('xyz')
     )
     assert_frame_equal(result, expected)
+
+
+def test_transpose_dt64tz():
```
Comment (mroeschke) [inline pandas/tests/frame/test_reshape.py]: Could you also test transposing a mixed dtype dataframe if we aren't already (e.g. datetimetz and int)?

Comment (jbrockmendel) [reply pandas/tests/frame/test_reshape.py]: Will do.  Only real difference is that it doesn't round-trip (though I think that behavior is correct)

Comment (mroeschke) [reply pandas/tests/frame/test_reshape.py]: Yeah that makes sense. Just good to ensure that there's no aggressive coercion of other data dtypes in the DataFrame once transposed i.e. the columns in `df.T` end up as object.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -683,11 +685,52 @@ def transpose(self, *args, **kwargs):
...
+            # transpose is a no-op, and passing axes would raise ValueError
+            #  as the DatetimeIndex.transpose method does not accept that kwarg
+            tz = values.tz
+            utc_values = values.asi8.reshape(self.shape)
+            utc_values = utc_values.transpose(axes_numbers)
+            if copy:
+                utc_values = utc_values.copy()
+            result = self._constructor(utc_values, **new_axes)
+            if self.ndim > 2:
+                # We're assuming DataFrame from here on
```
Comment (jreback) [inline pandas/core/generic.py]: not sure why you should do this here at all. Move this to internals and just create new blocks, rather than hacking it.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/generic.py]: The point is that it is going to be a hack regardless unless we allow 2D EAs.

Comment (jreback) [reply pandas/core/generic.py]: not going to happen any time soon if ever

as i said this is way too hacky and needs to be pushed to internals

**Comment by jreback** [comment]
@jbrockmendel I would rather you set this while transpose bizness aside. This needs to be pushed into internals, and does not need EA 2-D support to make this work. This kludge won't be accepted in its current form. Much rather get datetime EA fnished. This is not neceessary for that.

**Comment by TomAugspurger** [comment]
Will look closely later, but I think the transpose stuff will be handled automatically by internals when DatetimeArray is an extension array.

**Comment by jbrockmendel** [comment]
Travis fail is the test_rank segfault.

An argument against 2D DatetimeArray: `freq` wouldn't make much sense.

When other things clear up I'll take a look at pushing this into internals.  This bug is high on my list of pet peeves at the moment, so I'm not likely to let it go.

**Comment by jreback** [comment]
@jbrockmendel I would put this aside. This is just a hack and not worth the time / effort right now. Lots more fish to fry. This 'bug' has existed for quite a long time, and to be honest this is just a not-that interesting edge case.

**Comment by jbrockmendel** [comment]
Closing as not actionable.

**Pull request closed without merging**